_posts/2023-05-17-Scaling_part_2.md

```diff
@@ -0,0 +1,642 @@
+The key thing is to set the common name in the certificate to the name of the user we want to authenticate as.
+
+Since `kubeadm` always sets the same common name, we will here use `openssl` to generate the client certificates so that we get control over it.
+
+```bash
+# Generate etcd client certificate
+openssl req -newkey rsa:2048 -nodes -subj "/CN=test" \
+ -keyout "/tmp/test/pki/apiserver-etcd-client.key" -out "/tmp/test/pki/apiserver-etcd-client.csr"
+openssl x509 -req -in "/tmp/test/pki/apiserver-etcd-client.csr" \
+  -CA /tmp/test/pki/etcd/ca.crt -CAkey /tmp/test/pki/etcd/ca.key -CAcreateserial \
+  -out "/tmp/test/pki/apiserver-etcd-client.crt" -days 365
+```
+
+### Deploying the API server
+
+In order to deploy the API server, we will first need to generate some more certificates.
+The client certificates for connecting to etcd are already ready, but it also needs certificates to secure the exposed API itself, and a few other things.
+Then we will also need to create secrets from all of these certificates:
+
+```bash
+kubeadm init phase certs ca --config kubeadm-config.yaml
+kubeadm init phase certs apiserver --config kubeadm-config.yaml
+kubeadm init phase certs sa --cert-dir /tmp/test/pki
+
+kubectl create ns workload-api
+kubectl -n workload-api create secret tls test-ca --cert /tmp/test/pki/ca.crt --key /tmp/test/pki/ca.key
+kubectl -n workload-api create secret tls test-etcd --cert /tmp/test/pki/etcd/ca.crt --key /tmp/test/pki/etcd/ca.key
+kubectl -n workload-api create secret tls "test-apiserver-etcd-client" \
+  --cert "/tmp/test/pki/apiserver-etcd-client.crt" \
+  --key "/tmp/test/pki/apiserver-etcd-client.key"
+kubectl -n workload-api create secret tls apiserver \
+  --cert "/tmp/test/pki/apiserver.crt" \
+  --key "/tmp/test/pki/apiserver.key"
+kubectl -n workload-api create secret generic test-sa \
+  --from-file=tls.crt="/tmp/test/pki/sa.pub" \
+  --from-file=tls.key="/tmp/test/pki/sa.key"
+```
+
+With all that out of the way, we can finally deploy the API server!
+For this we will use a normal Deployment.
+
+```bash
+# Deploy API server
+curl -L https://github.com/Nordix/metal3-clusterapi-docs/raw/main/metal3-scaling-experiments/manifests/v2/kube-apiserver-deployment.yaml |
+  sed "s/CLUSTER/test/g" | kubectl -n workload-api apply -f -
+kubectl -n workload-api wait --for=condition=Available deploy/test-kube-apiserver
+```
+
+Time to check if it worked!
+We can use port-forwarding to access the API, but of course we will need some authentication method for it to be useful.
+With kubeadm we can generate a kubeconfig based on the certificates we already have.
+
+```bash
+kubeadm kubeconfig user --client-name kubernetes-admin --org system:masters \
+  --config tmp/kubeadm-config.yaml > kubeconfig.yaml
+```
+
+Now open another terminal and set up port-forwarding to the API server:
+
+```bash
+kubectl -n workload-api port-forward svc/test-kube-apiserver 6443
```

Comment:
So far I was using kubeadm and kubectl from inside kind container ? is what you expect from the reader or he should download these tools and set them in the host machine ? Is this set somewhere on top of this part or part 1? 
definitely at this point from inside the container this rise the error 
```
root@kind-control-plane:/# kubectl -n workload-api port-forward svc/test-kube-apiserver 6443
Unable to listen on port 6443: Listeners failed to create with the following errors: [unable to create listener: Error listen tcp4 127.0.0.1:6443: bind: address already in use unable to create listener: Error listen tcp6 [::1]:6443: bind: address already in use]
error: unable to listen on any of the requested ports: [{6443 6443}]
```

Reply:
I think kubeadm was meant to be running from the host machine. That may need to be stated explicitly; anw there was no instruction to run it inside a container either.

Reply:
Good point, thanks! This is all meant to be on the host. I think I mentioned it in part 1, but I will reiterate it here also

Reply:
Added